courses/intro-to-ada/contracts.rst

```diff
@@ -0,0 +1,485 @@
+
+       function Double (A : Int_8) return Int_8 is
+         (A * A)
+         with Post => Double'Result > A;
+
+       procedure Double (A : in out Int_8_Array)
+         with Post => (for all I in A'Range =>
+                         A (I) = A'Old (I) * A'Old (I))
+       is
+       begin
+          for V of A loop
+             V := Double (V);
+          end loop;
+       end Double;
+
+       V : Int_8_Array := (9, 10, 11);
+    begin
+       for E of V loop
+          Put_Line ("Original: " & Int_8'Image (E));
+       end loop;
+       New_Line;
+
+       Double (V);
+       for E of V loop
+          Put_Line ("Double:   " & Int_8'Image (E));
+       end loop;
+    end Show_Simple_Postcondition;
+
+In this example, we declare a signed 8-bit type :ada:`Int_8` and an array
+of that type (:ada:`Int_8_Array`). We want to ensure that, when calling
+the procedure :ada:`Double` for an object of :ada:`Int_8_Array` type, each
+element of the array will be doubled. This is implemented by a
+postcondition that uses a :ada:`for all` expression. The postcondition
+also makes use of the original value of the parameter before the call.
+The :ada:`'Old` attribute is used in this case to retrieve the original
+value.
+
+Also, we want to ensure that, in calls to the
+:ada:`Double` function for the :ada:`Int_8` type, the result will be
+greater than the input value. This is implemented by a postcondition that
+uses the :ada:`'Result` attribute of the function and compares it to the
+input value.
+
+We can use pre and postconditions at the same time in the declaration of
+a subprogram. For example:
+
+.. code:: ada
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO; use Ada.Text_IO;
+    with Ada.Numerics.Elementary_Functions;
+
+    procedure Show_Simple_Contract is
+       pragma Assertion_Policy (Pre  => Check,
+                                Post => Check);
+
+       type Int_8 is range -2 ** 7 .. 2 ** 7 - 1
+         with Size => 8;
+
+       function Double (A : Int_8) return Int_8 is
+         (A * A)
+         with
+              Pre  => (Integer'Size >= Int_8'Size * 2 and
+                       Integer (A) * Integer (A) < Integer (Int_8'Last)),
+              Post => Double'Result > A;
+
+       V : Int_8;
+    begin
+       V := Double (11);
+       Put_Line ("Double of 11 is " & Int_8'Image (V));
+
+       V := Double (12);   --  Precondition will fail...
+       Put_Line ("Double of 12 is " & Int_8'Image (V));
+    end Show_Simple_Contract;
+
+In this example, we want to ensure  that, in calls to the
+:ada:`Double` function for the :ada:`Int_8` type, the input value will not
+overflow in the call to the function. This is implemented by converting
+the input value to the :ada:`Integer` type, which is used to store the
+temporary calculation, and check if the result is still in the appropriate
+range for the :ada:`Int_8` type. The postcondition in this example is the
+same as in the previous example.
+
+Type invariants
```

Comment:
I think that teaching a concept by starting with a specific use-case and generalizing it is as valid as teaching it by starting with the general approach and mentioning specific use-cases. Also, for the examples added here, I have the impression that the order doesn't really matter, since they highlight similarities between predicates and type invariants.

In any case, although I'm not really convinced that it matters for this section, I'll change the order for you and start with predicates and then type invariants. :-)